test/integration/int_server.cc

```diff
@@ -0,0 +1,813 @@
+      // TODO return error to caller
+      return;
+    }
+
+    if (delay <= std::chrono::milliseconds(0)) {
+      ENVOY_LOG(debug, "ServerStream({}:{}:{}) sending response headers",
+                connection_.name(), connection_.id(), id_);
+      stream_encoder_.encodeHeaders(response_headers, true);
+      return;
+    }
+
+    // Limitation: at most one response can be sent on a stream at a time.
+    assert(nullptr == delay_timer.get());
+    if (delay_timer.get()) {
+      return;
+    }
+
+    response_headers_ =
+        std::make_unique<Envoy::Http::HeaderMapImpl>(response_headers);
+    delay_timer = connection_.dispatcher().createTimer([this, delay]() {
+      ENVOY_LOG(
+          debug,
+          "ServerStream({}:{}:{}) sending response headers after {} msec delay",
+          connection_.name(), connection_.id(), id_,
+          static_cast<long int>(delay.count()));
+      stream_encoder_.encodeHeaders(*response_headers_, true);
+      delay_timer->disableTimer();
+      delay_timer = nullptr;
+      response_headers_ = nullptr;
+    });
+    delay_timer->enableTimer(delay);
+  }
+
+  virtual void sendGrpcResponse(
+      Envoy::Grpc::Status::GrpcStatus status,
+      const Envoy::Protobuf::Message &message,
+      const std::chrono::milliseconds delay) override {
+    if (delay <= std::chrono::milliseconds(0)) {
+      ENVOY_LOG(debug, "ServerStream({}:{}:{}) sending gRPC response",
+                connection_.name(), connection_.id(), id_);
+      stream_encoder_.encodeHeaders(
+          Envoy::Http::TestHeaderMapImpl{{":status", "200"}}, false);
+      Envoy::Buffer::InstancePtr serialized_response =
+          Envoy::Grpc::Common::serializeBody(message);
+      stream_encoder_.encodeData(*serialized_response, false);
+      stream_encoder_.encodeTrailers(Envoy::Http::TestHeaderMapImpl{
+          {"grpc-status", std::to_string(static_cast<uint32_t>(status))}});
+      delete_after_callback_ = true;
+      return;
+    }
+
+    // Limitation: at most one response can be sent on a stream at a time.
+    assert(nullptr == delay_timer.get());
+    if (delay_timer.get()) {
+      return;
+    }
+
+    delete_after_callback_ = false;
+    response_status_ = status;
+    response_body_ = Envoy::Grpc::Common::serializeBody(message);
+    delay_timer = connection_.dispatcher().createTimer([this, delay]() {
+      ENVOY_LOG(
+          debug,
+          "ServerStream({}:{}:{}) sending gRPC response after {} msec delay",
+          connection_.name(), connection_.id(), id_,
+          static_cast<long int>(delay.count()));
+      stream_encoder_.encodeHeaders(
+          Envoy::Http::TestHeaderMapImpl{{":status", "200"}}, false);
+      stream_encoder_.encodeData(*response_body_, false);
+      stream_encoder_.encodeTrailers(Envoy::Http::TestHeaderMapImpl{
+          {"grpc-status",
+           std::to_string(static_cast<uint32_t>(response_status_))}});
+      delay_timer->disableTimer();
+      connection_.removeStream(id_);
```

Comment:
Why do you not need to call the request_callback_ here (i.e. set delete_after_callback) ?